packages/kbn-content-management-utils/src/saved_object_content_storage.ts

```diff
@@ -0,0 +1,396 @@
+    const {
+      saved_object: savedObject,
+      alias_purpose: aliasPurpose,
+      alias_target_id: aliasTargetId,
+      outcome,
+    } = await soClient.resolve<Types['Attributes']>(this.savedObjectType, id);
+
+    const response: Types['GetOut'] = {
+      item: savedObjectToMapItem(savedObject, false),
+      meta: {
+        aliasPurpose,
+        aliasTargetId,
+        outcome,
+      },
+    };
+
+    // Validate DB response and DOWN transform to the request version
+    const { value, error: resultError } = transforms.get.out.result.down<
+      Types['GetOut'],
+      Types['GetOut']
+    >(response);
+
+    if (resultError) {
+      throw Boom.badRequest(`Invalid response. ${resultError.message}`);
+    }
+
+    return value;
+  }
+
+  async bulkGet(): Promise<never> {
+    // Not implemented
+    throw new Error(`[bulkGet] has not been implemented. See ${this.constructor.name} class.`);
+  }
+
+  async create(
+    ctx: StorageContext,
+    data: Types['Attributes'],
+    options: Types['CreateOptions']
+  ): Promise<Types['CreateOut']> {
+    const transforms = ctx.utils.getTransforms(this.cmServicesDefinition);
+    const soClient = await savedObjectClientFromRequest(ctx);
+
+    // Validate input (data & options) & UP transform them to the latest version
+    const { value: dataToLatest, error: dataError } = transforms.create.in.data.up<
+      Types['Attributes'],
+      Types['Attributes']
+    >(data);
+    if (dataError) {
+      throw Boom.badRequest(`Invalid data. ${dataError.message}`);
+    }
+
+    const { value: optionsToLatest, error: optionsError } = transforms.create.in.options.up<
+      Types['CreateOptions'],
+      Types['CreateOptions']
+    >(options);
+    if (optionsError) {
+      throw Boom.badRequest(`Invalid options. ${optionsError.message}`);
+    }
+
+    const createOptions = this.createArgsToSoCreateOptions(optionsToLatest);
+
+    // Save data in DB
+    const savedObject = await soClient.create<Types['Attributes']>(
+      this.savedObjectType,
+      dataToLatest,
+      createOptions
+    );
+
+    // Validate DB response and DOWN transform to the request version
+    const { value, error: resultError } = transforms.create.out.result.down<
+      Types['CreateOut'],
+      Types['CreateOut']
+    >({
+      item: savedObjectToMapItem(savedObject, false),
+    });
+
+    if (resultError) {
+      throw Boom.badRequest(`Invalid response. ${resultError.message}`);
+    }
+
+    return value;
+  }
+
+  async update(
+    ctx: StorageContext,
+    id: string,
+    data: Types['Attributes'],
+    options: Types['UpdateOptions']
+  ): Promise<Types['UpdateOut']> {
+    const transforms = ctx.utils.getTransforms(this.cmServicesDefinition);
+    const soClient = await savedObjectClientFromRequest(ctx);
+
+    // Validate input (data & options) & UP transform them to the latest version
+    const { value: dataToLatest, error: dataError } = transforms.update.in.data.up<
+      Types['Attributes'],
```

Comment:
@sebelga All these transforms take the same argument twice. I bet I'm overlooking something. Can you tell me what?

Reply:
It's the IN and OUT. At this stage we don't know what's the out interface (it could be V2, V3, V4 of the attributes) so to make TS happy we declare the same IN and OUT.

Looking at it, The OUT type could probably implicitly be equal to the IN to avoid this duplication.

Reply:
Now that I look at it again, it really does not matter here as it is all abstracted away.